File: apps/scala/design-pattern/test/DecoratorTest.scala
package designpattern.decorator

class DecoratorSuite extends munit.FunSuite:
  test("SimpleWriter はそのまま出力する") {
    val writer = SimpleWriter()
    assertEquals(writer.writeLine("Hello"), "Hello")
  }

  test("NumberingWriter で行番号を付与する") {
    val writer = NumberingWriter(SimpleWriter())
    assertEquals(writer.writeLine("Hello"), "1: Hello")
    assertEquals(writer.writeLine("World"), "2: World")
  }

  test("TimestampWriter でタイムスタンプを付与する") {
    val writer = TimestampWriter(SimpleWriter(), "2024-01-01")
    assertEquals(writer.writeLine("Hello"), "[2024-01-01] Hello")
  }

  test("デコレータを積み重ねる") {
    val writer = TimestampWriter(
      NumberingWriter(SimpleWriter()),
      "2024-01-01"
    )
    assertEquals(writer.writeLine("Hello"), "[2024-01-01] 1: Hello")
  }

  test("CheckSumWriter でチェックサムを付与する") {
    val writer = CheckSumWriter(SimpleWriter())
    val output = writer.writeLine("Hello")
    assert(output.startsWith("Hello [checksum:"))
  }

  test("writeLines で複数行を処理する") {
    val writer = NumberingWriter(SimpleWriter())
    val lines = writer.writeLines(Seq("A", "B", "C"))
    assertEquals(lines, Seq("1: A", "2: B", "3: C"))
  }

File: apps/java/design-pattern/src/test/java/pattern/decorator/DecoratorTest.java
package pattern.decorator;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class DecoratorTest {

    @Test
    void simpleWriterWritesLines() {
        SimpleWriter writer = new SimpleWriter();

        writer.writeLine("Hello");
        writer.writeLine("World");
        writer.close();

        assertEquals("Hello\nWorld\n", writer.getContents());
    }

    @Test
    void numberingWriterAddsLineNumbers() {
        SimpleWriter simple = new SimpleWriter();
        Writer writer = new NumberingWriter(simple);

        writer.writeLine("First");
        writer.writeLine("Second");
        writer.close();

        String contents = simple.getContents();
        assertTrue(contents.contains("1: First"));
        assertTrue(contents.contains("2: Second"));
    }

    @Test
    void timeStampingWriterAddsTimestamp() {
        SimpleWriter simple = new SimpleWriter();
        Writer writer = new TimeStampingWriter(simple);

        writer.writeLine("Event");
        writer.close();

        String contents = simple.getContents();
        assertTrue(contents.contains("Event"));
        // Timestamp format contains colons from time
        assertTrue(contents.matches("(?s).*\\d{4}-\\d{2}-\\d{2}.*Event.*"));
    }

    @Test
    void decoratorsCanBeStacked() {
        SimpleWriter simple = new SimpleWriter();
        Writer writer = new NumberingWriter(new TimeStampingWriter(simple));

        writer.writeLine("Line A");
        writer.writeLine("Line B");
        writer.close();

        String contents = simple.getContents();
        assertTrue(contents.contains("1: "));
        assertTrue(contents.contains("2: "));
        assertTrue(contents.contains("Line A"));
        assertTrue(contents.contains("Line B"));
    }

    @Test
    void writerDecoratorDelegatesToWrapped() {
        SimpleWriter simple = new SimpleWriter();
        WriterDecorator decorator = new WriterDecorator(simple);

        decorator.writeLine("delegated");
        decorator.close();

        assertEquals("delegated\n", simple.getContents());
    }
}
